### PR TITLE
EN-11545: Add transform from column names to ids

### PR DESCRIPTION
**Pull request by aerust**
- Add transform(.) function to be used in core
   for transforming a strategy from using column
   names to column ids.
 - Also add augment(.) functions for GeoRegion
   computation strategies.

### CODE REVIEW

**Comment by chitang** [comment]
Would it read better to use comprehension like?

for {
         a <- Right("a").right
         b <- Left("b").right
         c <- Right("c").right
         d <- Left("d").right
      } yield {
        Right(a + b + c).right
      }

**Comment by aerust** [comment]
Where exactly? Or just generally?

File: src/main/scala/com/socrata/computation_strategies/GeocodingComputationStrategy.scala [hunk 1]
```diff
@@ -212,4 +212,35 @@ object GeocodingComputationStrategy extends ComputationStrategy with Augment[Fle
       }
     case None => None
   }
+
+  case class UnknownSourceColumnException[CN](cn: CN) extends Exception
+  override protected def transform[CN: JsonDecode, CI : JsonEncode](parameters: JObject, columns: Map[CN, CI]):
+    Either[ValidationError, JObject] = {
+      def map(cn: Option[CN]): Option[CI] = cn.map { cn =>
```
Comment (chitang) [inline src/main/scala/com/socrata/computation_strategies/GeocodingComputationStrategy.scala]: how about rename map to toCI or ci

**Comment by chitang** [comment]
At least in ComputationStrategy.transform where Either/return are used.

File: src/main/scala/com/socrata/computation_strategies/ComputationStrategy.scala [hunk 2]
```diff
@@ -65,6 +65,21 @@ trait ComputationStrategy {
 
   protected def validate[ColumnName : JsonDecode](definition: StrategyDefinition[ColumnName],
                                                   columns: Option[Map[ColumnName, SoQLType]]): Option[ValidationError]
+
+  def transform[CN: JsonDecode, CI : JsonEncode](definition: StrategyDefinition[CN], columns: Map[CN, CI]):
```
Comment (chitang) [inline src/main/scala/com/socrata/computation_strategies/ComputationStrategy.scala]: remove space before colon in "CI : JsonEncode"

File: src/main/scala/com/socrata/computation_strategies/ComputationStrategy.scala [hunk 3]
```diff
@@ -65,6 +65,21 @@ trait ComputationStrategy {
...
+    Either[ValidationError, StrategyDefinition[CI]] = {
+      import EitherUtil._
+
+      for {
+        sourceColumns <- either(definition.sourceColumns.map(mapOrLeft(_, columns, UnknownSourceColumn(_ : CN)))).right
+        parameters <- either(definition.parameters.map(transform(_, columns))).right
+      } yield {
+        return Right(StrategyDefinition(definition.typ, sourceColumns, parameters))
```
Comment (chitang) [inline src/main/scala/com/socrata/computation_strategies/ComputationStrategy.scala]: remove return

Comment (chitang) [inline src/main/scala/com/socrata/computation_strategies/ComputationStrategy.scala]: Remove both return and Right().

Comment (aerust) [inline src/main/scala/com/socrata/computation_strategies/ComputationStrategy.scala]: oh I should remove the Right(.) too

File: src/main/scala/com/socrata/computation_strategies/ComputationStrategy.scala [hunk 4]
```diff
@@ -83,14 +98,33 @@ object ComputationStrategy {
...
 
 trait Augment[T] {
-
   def augment[ColumnName : JsonDecode : JsonEncode](definition: StrategyDefinition[ColumnName],
                                                    info: T): Either[ValidationError, StrategyDefinition[ColumnName]]
 }
+
+object EitherUtil {
+
+  def orLeft[T, U, V](t: T, map: Map[T, U], orElse: T => V): Either[V, U] =
```
Comment (chitang) [inline src/main/scala/com/socrata/computation_strategies/ComputationStrategy.scala]: This is used only once.  Is this the right replacement?

        cn match {
          case None => Right(None)
          case Some(c) =>
            columns.get(c).map(ci => Right(Some(ci))).getOrElse(Left(UnknownSourceColumn(c)))
        }

Let's see if you still need all these option/either/map utils after you separate class definition from json codec.

File: src/main/scala/com/socrata/computation_strategies/ComputationStrategy.scala [hunk 5]
```diff
@@ -83,14 +98,33 @@ object ComputationStrategy {
...
+  def transform[CN : JsonDecode, CI : JsonEncode](definition: StrategyDefinition[CN],
+                                     columns: Map[CN, CI]): Either[ValidationError, StrategyDefinition[CI]] =
+    strategies(definition.typ).transform(definition, columns)
 }
 
 trait ParameterSchema {
   def requiredFields: Seq[String]
 }
 
 trait Augment[T] {
```
Comment (chitang) [inline src/main/scala/com/socrata/computation_strategies/ComputationStrategy.scala]: Let's have a more specific name than augment.  Again, not sure if we still need it if you separate class definition and json codec.

**Comment by aerust** [comment]
Chi says I can merge and come back to addressing JObject stuff later.

**Pull request merged**